python/ray/data/tests/test_dataset_image.py

```diff
@@ -1,193 +1,142 @@
-    assert sorted(ds.to_pandas()["label"]) == ["cat", "cat", "dog"]
-
-
-def test_image_folder_datasource_size_parameter(
-    ray_start_regular_shared, enable_automatic_tensor_extension_cast
-):
-    """Test `ImageFolderDatasource` size parameter works with differently-sized images.
-
-    The folder "different-sizes" contains two cat images and one dog image. Each image
-    has a different size, with the size described in file names (e.g., 32x32.png). All
-    images are RGB images.
-    """
-    root = "example://image-folders/different-sizes"
-    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root, size=(32, 32))
-
-    tensors = ds.to_pandas()["image"]
-    assert all(tensor.shape == (32, 32, 3) for tensor in tensors)
-
-
-def test_image_folder_datasource_retains_shape_without_cast(
-    ray_start_regular_shared, enable_automatic_tensor_extension_cast
-):
-    """Test `ImageFolderDatasource` retains image shapes if casting is disabled.
-
-    The folder "different-sizes" contains two cat images and one dog image. The image
-    sizes are 16x16, 32x32, and 64x32. All images are RGB images.
-    """
-    if enable_automatic_tensor_extension_cast:
-        return
-
-    root = "example://image-folders/different-sizes"
-    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root)
-    arrays = ds.to_pandas()["image"]
-    shapes = sorted(array.shape for array in arrays)
-    assert shapes == [(16, 16, 3), (32, 32, 3), (64, 64, 3)]
-
-
-@pytest.mark.parametrize(
-    "mode, expected_shape", [("L", (32, 32)), ("RGB", (32, 32, 3))]
-)
-def test_image_folder_datasource_mode_parameter(
-    mode,
-    expected_shape,
-    ray_start_regular_shared,
-    enable_automatic_tensor_extension_cast,
-):
-    """Test `ImageFolderDatasource` works with images from different colorspaces.
-
-    The folder "different-modes" contains two cat images and one dog image. Their modes
-    are "CMYK", "L", and "RGB". All images are 32x32.
-    """
-    root = "example://image-folders/different-modes"
-    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root, mode=mode)
-
-    tensors = ds.to_pandas()["image"]
-    assert all([tensor.shape == expected_shape for tensor in tensors])
-
-
-@pytest.mark.parametrize("size", [(-32, 32), (32, -32), (-32, -32)])
-def test_image_folder_datasource_value_error(ray_start_regular_shared, size):
-    root = "example://image-folders/simple"
-    with pytest.raises(ValueError):
-        ray.data.read_datasource(ImageFolderDatasource(), root=root, size=size)
-
-
-def test_image_folder_datasource_e2e(shutdown_only):
-    from ray.train.torch import TorchCheckpoint, TorchPredictor
-    from ray.train.batch_predictor import BatchPredictor
-
-    from torchvision import transforms
-    from torchvision.models import resnet18
-
-    root = "example://image-folders/simple"
-    dataset = ray.data.read_datasource(
-        ImageFolderDatasource(), root=root, size=(32, 32)
+class TestReadImages:
```

Comment:
Any reason for making this a test class rather than a flat set of module-level functions? Usually this is done if there's some shared setup or state.

Reply:
Couple reasons:
* Better organization. `test_dataset_formats` uses a flat structure, and related tests are often far apart.  For example, `read_text` and `read_text_remote_args` were 1500 lines apart. By grouping `read_images` tests in a class, we ensure that `read_images`-related tests are located close to each other. 
* Shorter test names.`test_invalid_size` is easier to read than `test_read_images_invalid_size`

Also, we do something similar with the checkpoint tests, so there's atleast some precedence for this:
https://github.com/ray-project/ray/blob/2217f0ca9d3aa72c6f12fe79e3fd0d502ef527c8/python/ray/air/tests/test_checkpoints.py#L159

Reply:
> In any case, we might add non-ray.data.read_images() test to this module

Yeah, this is what I'm concerned about. 

If we add non-`ray.data.read_images` tests, I doubt we're going to go back and group the `read_images` tests in a class. Whereas, if we start with a class grouping, I figure people will pattern-match the existing tests and group new tests under a class. This should avoid the `dataset_formats` situation where we have many disorganized tests.

> For a general guide, I'd be in favor of something like:
> * source module --> test module
> * source abstraction (function or class) --> test module if single abstraction in module, otherwise test class
> * source class method --> test class (pytest supports nested test classes AFAIK)

I agree with this except for the "test module if single abstraction in module". I think in practice people will always pattern match existing tests, so if we start with a flat structure people will continue to use that (even if we test unrelated abstractions).